### PR TITLE
Add `groupby.transform` (only support for aggregations)

### PR DESCRIPTION
**Pull request by shwina**
Closes https://github.com/rapidsai/cudf/issues/4522

This PR adds support for doing groupby aggregations via the `transform()` API, where the result of the aggregation is broadcasted to the size of the group. Note that more general transformations are not supported at this time.

### CODE REVIEW

File: docs/cudf/source/basics/groupby.rst [hunk 1]
```diff
@@ -220,6 +220,27 @@ Limitations
  .. |describe| replace:: ``describe``
  .. _describe: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#flexible-apply
 
+
+ Transform
+---------
+
+The `.transform()` method aggregates per group, and broadcasts the
```
Comment (shwina) [inline docs/cudf/source/basics/groupby.rst:227]:
```suggestion
The ``.transform()`` method aggregates per group, and broadcasts the
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 2]
```diff
@@ -655,6 +655,51 @@ def rolling_avg(val, avg):
...
+             b
+          0  5
+          1  3
+          2  3
+          3  5
+          4  5
+
+        """
+        try:
+            # first, try aggregating:
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:689]:
```suggestion
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 3]
```diff
@@ -655,6 +655,50 @@ def rolling_avg(val, avg):
...
         return grouped_values.apply_chunks(function, **kwargs)
 
+    def transform(self, function):
+        """Apply an aggregation, then broadcast the result to the group size.
+
+        Parameters
+        ----------
+        function: str or callable
+            Aggregation to apply to each group. Note that currently,
+            only aggregations are supported by `transform`.
```
Comment (isVoid) [inline python/cudf/cudf/core/groupby/groupby.py:665]: I pulled this locally and tried several inputs. It seems for callables only form `lambda x: x.agg` is accepted, should we make it clear in the docstring?

Comment (shwina) [reply python/cudf/cudf/core/groupby/groupby.py]: I added a note that hopefully clarifies this. Let me know if you would add/rephrase anything.

Comment (isVoid) [reply python/cudf/cudf/core/groupby/groupby.py]: So currently it redirects the user to look up for `agg`, and that's ok. I think there are two things we can improve: we can add a `see also` section here for easy redirection. But more importantly in `agg` docstring the description for `func` should be clearer. The latter is out of the scope of this PR, I think.

**Review by isVoid — APPROVED** [review]
Thanks!

**Comment by shwina** [comment]
rerun tests

File: docs/cudf/source/basics/groupby.rst [hunk 4]
```diff
@@ -220,6 +222,27 @@ Limitations
  .. |describe| replace:: ``describe``
  .. _describe: https://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html#flexible-apply
 
+
+ Transform
+---------
```
Comment (shwina) [inline docs/cudf/source/basics/groupby.rst:227]:
```suggestion
Transform
---------
```

File: python/cudf/cudf/core/groupby/groupby.py [hunk 5]
```diff
@@ -184,11 +184,25 @@ def agg(self, func):
         Parameters
         ----------
         func : str, callable, list or dict
+            Argument specifying the aggregation(s) to perform on the
+            groups. `func` can be any of the following types:
```
Comment (shwina) [inline python/cudf/cudf/core/groupby/groupby.py:188]:
```suggestion
            groups. `func` can be any of the following:
```

**Comment by shwina** [comment]
rerun tests

**Review by galipremsagar — APPROVED** [review]

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
rerun tests

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**